### PR TITLE
Warn when objectivec compiled with ARC

### PR DESCRIPTION
**Pull request by mjmdavis**
Add an error and an explain how to fix it when objectivec generated files are compiled with ARC.

This could save some people a bit of time working out how to use this in their ARC enabled projects.

### CODE REVIEW

**Comment by mjmdavis** [comment]
Signed.

**Comment by mjmdavis** [comment]
When compiling you get an error if ARC is enabled for the generated file and a hint on how to disable ARC for the file.

File: src/google/protobuf/compiler/objectivec/objectivec_file.cc [hunk 1]
```diff
@@ -534,6 +534,12 @@ void FileGenerator::PrintFileRuntimePreamble(
       "\n",
       "filename", file_->name());
 
+  printer->Print(
+    "#if __has_feature(objc_arc)\n"
+    "  #error This file should be compiled with ARC. Use the -fobjc-arc Compiler Flag to disable ARC for this file.\n"
+    "#endif"
+    "\n");
+
```
Comment (mjmdavis) [inline src/google/protobuf/compiler/objectivec/objectivec_file.cc]: The string here is wrong, should say: "This file should **not** be compiled with ARC...".
This warning is also included in the header file which it should not be. :/

File: src/google/protobuf/compiler/objectivec/objectivec_file.cc [hunk 2]
```diff
@@ -310,6 +310,12 @@ void FileGenerator::GenerateSource(io::Printer *printer) {
   vector<const FileDescriptor*> deps_with_extensions;
   CollectMinimalFileDepsContainingExtensions(file_, &deps_with_extensions);
 
+  printer->Print(
+    "#if __has_feature(objc_arc)\n"
+    " #error This file should not be compiled with ARC. Use the -fobjc-arc Compiler Flag to disable ARC for this file.\n"
```
Comment (thomasvl) [inline src/google/protobuf/compiler/objectivec/objectivec_file.cc]: Still wrong, you want -fno-objc-arc.

https://github.com/google/protobuf/tree/master/objectivec#building

Comment (mjmdavis) [inline src/google/protobuf/compiler/objectivec/objectivec_file.cc]: Seems I doubled down on the copy paste error 😳.

**Comment by mjmdavis** [comment]
@thomasvl I did use more than 1 email. I'm waiting on something to resolve that.

**Comment by mjmdavis** [comment]
Sorry about the churn on this.

**Comment by mjmdavis** [comment]
Signed

**Comment by mjmdavis** [comment]
Signed.

**Comment by mjmdavis** [comment]
I signed it!

**Comment by thomasvl** [comment]
Looking at the [patch](https://patch-diff.githubusercontent.com/raw/google/protobuf/pull/2118.patch) for all these changes; it shows two addresses.  The second one (not the gmail one), doesn't seem to be listed as having signed; and the company from that domain isn't listed either as having signed.

**Comment by mjmdavis** [comment]
@thomasvl we're working on getting it signed. For some reason the OSS contributions guy couldn't sign the document.

**Comment by mjmdavis** [comment]
We're in touch with Google CLA Admin to clear this up.

**Pull request closed without merging**